### PR TITLE
Add setting per-autolink for allowing bot posts

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

Currently, the plugin does not edit any posts made by bots. There is a customer wanting to use the autolink plugin with their own bot's posts. I've added a setting on each configured autolink to allow bot posts to be edited.

Thread for customer request:
https://community-daily.mattermost.com/core/pl/35ho6xwqepfppyb6wpx5fwqz9a

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-autolink/issues/141

#### Related Issues/PRs

Issue - https://github.com/mattermost/mattermost-plugin-autolink/issues/94
Pull Request - https://github.com/mattermost/mattermost-plugin-autolink/pull/98

### CODE REVIEW

File: plugin.json [hunk 1]
```diff
@@ -30,6 +30,12 @@
         "type": "bool",
         "default": false
       },
+      {
+        "key": "ProcessPostsFromBots",
+        "display_name": "Process Posts from Bots:",
+        "type": "bool",
+        "help_text": "When enabled, the plugin will be able to edit posts made by bots."
+      },
```
Comment (hanzei) [inline plugin.json:38]: Should this be a per-link setting?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: plugin.json (hunk 1 above)
Comment (aaronrothschild) [reply plugin.json:38]: I agree this could be a per link setting. But grasping for uses case examples....

**Review by aaronrothschild — APPROVED** [review]

**Review by levb — APPROVED** [review]

**Review by hanzei — APPROVED** [review]
If @aaronrothschild is fine with this, I'm also

**Comment by mickmister** [comment]
@hanzei Most of the consensus around blocking bot posts in the first place was that this should be a per-link setting, and I didn't want to force an admin to make a hard decision of global enable/disable.

**Review by hanzei — APPROVED** [review]
LGTM, but tests are broken https://app.circleci.com/pipelines/github/mattermost/mattermost-plugin-autolink/523/workflows/e0b365cc-f37e-4501-837d-e495617448ec/jobs/1629

**Comment by mickmister** [comment]
@hanzei The specific test failing is failing because it defends against fetching the bot user until we have determined that the text has changed by any one of the autolinks. I see two paths here:

- Don't fetch the bot user beforehand. Go through all autolinks like normal. If the text has been altered, run through the autolinks again with the if statement check added in the last commit.
- Take the RPC hit of fetching the user (I'd rather not)

**Comment by hanzei** [comment]
@mickmister Can we combine the two things and only fetch the bot once in the loop? Something like:

```

var user *model.user

for _, link := range conf.Links {
	avoidBotPost := false
	if !link.ProcessBotPosts {
		if user == nil {
			// fetch user
		}
		if user.IsBot() {
			avoidBotPost = true
		}
	}
	if !avoidBotPost && p.inScope(link.Scope, channelName, teamName) {
		newText = link.Replace(newText)
	}
}
```

**Comment by mickmister** [comment]
@hanzei By default all links will have `ProcessBotPosts` false, so the snippet you posted is not really solving the performance issue. What about this?

```go
var user *model.user

for _, link := range conf.Links {
    if !p.inScope(link.Scope, channelName, teamName) {
        continue
    }

    processed := link.Replace(newText)
    if processed == newText {
        continue
    }
    
    if !link.ProcessBotPosts {
        if user == nil {
            // fetch user
        }
        if user.IsBot() {
            continue
        }
    }

    newText = processed
}
```

**Comment by hanzei** [comment]
@mickmister `ProcessBotPosts` being `false` by default is a breaking change right? Previously, all bot posts where shipped.

**Comment by mickmister** [comment]
> `ProcessBotPosts` being `false` by default is a breaking change right? Previously, all bot posts where shipped

@hanzei No, the default behavior has not changed in this PR. I think the confusion may be around the `ProcessBotPosts` boolean. If it is false, we will skip the bot post like usual. "Process" in this case means "accept" the bot posts for edits.

Before the PR:
- If no autolinks edited the post, do not check if the user is a bot
- If at least one autolink edited the post, check if the user is a bot.
    - If the user is a bot, return nil and do not persist any edits
    - else persist the changes

After this PR (let's just assume the default setting to compare with previous behavior):
- If a given autolink edited the post, check if the user is a bot
    - If the user is a bot, do not persist the change
    - else queue change for persisting
- Check if changes have been applied, and persist if changed

File: server/autolinkplugin/plugin.go [hunk 2]
```diff
@@ -181,9 +174,29 @@ func (p *Plugin) ProcessPost(c *plugin.Context, post *model.Post) (*model.Post,
...
+				out := link.Replace(newText)
+				if out == newText {
+					continue
+				}
+
+				if !link.ProcessBotPosts {
+					if author == nil && authorErr == nil {
+						author, authorErr = p.API.GetUser(post.UserId)
+						if authorErr != nil {
+							p.API.LogError("failed to check if message for rewriting was send by a bot", "error", authorErr)
```
Comment (hanzei) [inline server/autolinkplugin/plugin.go:190]: Should we copy the deleted comment for below:
```suggestion
							// NOTE: Not sure how we want to handle errors here, we can either:
							// * assume that occasional rewrites of Bot messges are ok
							// * assume that occasional not rewriting of all messages is ok
							// Let's assume for now that former is a lesser evil and carry on.
							p.API.LogError("failed to check if message for rewriting was send by a bot", "error", authorErr)
```

**Review by hanzei — APPROVED** [review]
LGTM :+1:

**Comment by hanzei** [comment]
@DHaussermann I would like to merge this PR soon as it's blocking the next release. Could the Deepak look into this one?

**Comment by dipak-demansol** [comment]
> @dipak-demansol Are you able to take a look at this PR? Let me know if you have any questions

sure i'll check it.

**Comment by DHaussermann** [comment]
@mickmister could there be some gaps autocomplete functionality.
- Can the command be included in the autocomplete?
- Can the command show an output that includes the value? 
- The help text should show that you can pass in the configuration for bots.

**Comment by iprok** [comment]
Hello!

I represent customer of mattermost too. And we are also really interested in this functionality. Can you approximately tell us when it will be released?

**Comment by DHaussermann** [comment]
@mickmister do you have any thoughts on @vinod-demansol's comments above. 

We should also add info to the help on how the new functionality is set.

**Comment by mickmister** [comment]
> - Can the command be included in the autocomplete?
> - Can the `/autolink list` command show an output that includes the value?
> - The help text should show that you can pass in the configuration for bots.

@DHaussermann I'm not sure when I will have the cycles to address the issues above. @aaronrothschild Are these items a requirement for releasing this feature? Maybe we can assign this PR to someone from Demansol?

> also I see if I type /autolink set mm ProcessBotPosts false , then also I am able to edit Bot message , So this is issue Please fix this issue.

@vinod-demansol Can you please explain what you mean by "I am able to edit Bot message"? What steps did you take to test (the "editing" part of the test), what was the expected outcome, and what happened instead?

**Comment by aaronrothschild** [comment]
@mickmister Let's try to ship it as is, those missing commands aren't blockers, but should be in the next release fo Autolink.  It seems we have some customers waiting for this, so I am inclined to ship as is.  Then add a new ticket for 1.4.0.

**Comment by hanzei** [comment]
@vinod-demansol What do you think about the comment above? https://github.com/mattermost/mattermost-plugin-autolink/pull/160#issuecomment-1021480299

**Comment by vinod-demansol** [comment]
> @vinod-demansol What do you think about the comment above? [#160 (comment)](https://github.com/mattermost/mattermost-plugin-autolink/pull/160#issuecomment-1021480299)

I am ok if its tracked in new ticket in 1.4.0

**Comment by vinod-demansol** [comment]
> /autolink set mm ProcessBotPosts false

Steps i followed 
1. Create a link 
2. create a pattern and template 

> > * Can the command be included in the autocomplete?
> > * Can the `/autolink list` command show an output that includes the value?
> > * The help text should show that you can pass in the configuration for bots.
> 
> @DHaussermann I'm not sure when I will have the cycles to address the issues above. @aaronrothschild Are these items a requirement for releasing this feature? Maybe we can assign this PR to someone from Demansol?
> 
> > also I see if I type /autolink set mm ProcessBotPosts false , then also I am able to edit Bot message , So this is issue Please fix this issue.
> 
> @vinod-demansol Can you please explain what you mean by "I am able to edit Bot message"? What steps did you take to test (the "editing" part of the test), what was the expected outcome, and what happened instead?

steps I followed 
1. created a link 
2. created a pattern and template 
"Pattern": "(LHS)",
"Template": "[LHS](https://docs.mattermost.com/process/training.html#lhs)"
3.Enabled the link
4. typed /autolink set mm ProcessBotPosts true
5. typed LHS in channel and I got the expected LHS template 
6. typed /autolink set mm ProcessBotPosts false
7. tried editing the previous typed LHS , then also I see template string

**Comment by hanzei** [comment]
@vinod-demansol What is the difference between the expected behaviour and the one that you are seeing?

**Comment by dipak-demansol** [comment]
curl -i -X POST -H 'Content-Type: application/json' -d '{"channel_id":<>, "message":"This is a message from a bot with mention for @dkh-sysadmin and @dkh-member", "props":{"attachments": [{"pretext": "LHS","text": "LHS"}]}}' -H 'Authorization: Bearer <>' https://demansol-sandbox.test.mattermost.cloud/api/v4/posts

@mickmister when i make a bot post using this curl script, i don't see the post re-written by autolink in the mattermost. can you pls provide a valid curl script in case its a user error.

**Comment by hanzei** [comment]
@dipak-demansol What part of the message do you expect to be re-written?

**Comment by hanzei** [comment]
@dipak-demansol What part of the message do you expect to be re-written?

**Comment by hanzei** [comment]
@dipak-demansol What part of the message do you expect to be re-written? 

FYI: Slack attachments where never meant to be re-written.

**Comment by mickmister** [comment]
@dipak-demansol I agree with @hanzei, it seems you are expecting `LHS` to be autolinked inside of the slack attachments?

If that's the case, we should indeed expect the text to _not_ change, so what you're seeing is not a bug in this case.

**Comment by dipak-demansol** [comment]
@mickmister can you pls provide me the repro steps to regenerate the issue.

**Comment by dipak-demansol** [comment]
> @dipak-demansol I agree with @hanzei, it seems you are expecting `LHS` to be autolinked inside of the slack attachments?
> 
> If that's the case, we should indeed expect the text to _not_ change, so what you're seeing is not a bug in this case.

yes i was expecting that. i misunderstood the issue so can you pls provide me the repro steps to generate the issue and then i'll test it properly.

**Comment by mickmister** [comment]
> * The Setting should show in help
> * The setting should be configurable with a UI command
> * The state of the setting should be returned when you run list?

@dipak-demansol These issues are now addressed

- Setting shows in help now
- `/autolink set (name) ProcessBotPosts true` shows up in autocomplete now. The command was working already before, just no autocomplete.
- `ProcessBotPosts` shows when the autolink has this value as `true`. This was actually working as well, it just only shows the value if it is `true`. This is the same as other boolean flags for the autolinks.

**Comment by dipak-demansol** [comment]
With master build when i add Bot post then LHS is not rewritten and its expected, if i try to set **ProcessBotPosts** True in old build then i'll get this kind of response.
 
image

:- With the PR build i'm able to set true the **ProcessBotPosts** and Now the bot post is rewritten so its **LGTM.**
image

non blocking point
-------------------------------
:- testing with PR build i'm able to **ProcessBotPosts** set True/false but it would be great if the True/false also be the part of auto-suggestion.
-------------------------------
@mickmister **After Discussing the same result with Dylan, if is there anything blocking then will update you, if all Good then i'll add tag as Qa  review Done**

**Review by dipak-demansol — APPROVED** [review]
**LGTM**, as per the Discussion with Dylan, will add non-blocking point list on new tiket.

**Comment by dipak-demansol** [comment]
> **LGTM**, as per the Discussion with Dylan, will add non-blocking point list on new tiket.

@hanzei @mickmister

**Pull request merged**